File: 03 - Probability/golang/03_probability_solution.go
package main

import (
	"fmt"
	"math/rand"
)

// normalizeMap takes in map of floats and normalized it
// so tha that they sum up to one
func normalizeMap(m map[string]float64) {

	// accumulator variable
	var sum float64

	// calculate the sum of the given floats
	for _, v := range m {
		sum += v
	}

	// divide total on each float to create a sort of
	// probability distribution
	for k, v := range m {
		m[k] = v / sum
	}
}

// choose function takes in a probability distribution
// and returns one based on a uniformly selected random number
func choose(m map[string]float64) string {

	// uniform random number between o.o and 1.0
	r := rand.Float64()

	// key to store the value to return
	var key string

	// iterate over the map
	for k, v := range m {

		// check r and update its value, when greater than 0
		if r >= 0.0 {
			r -= v

			// update key with the current value of k
			key = k

		} else {

			// break out of the loop when r goes below 0
			break
		}
	}

	// return the value that helped cross the r threshold
	return key
}

// simulateThrows simulates a coin throw for given number of times,
// map in the input represents the probability distribution of the sides
func simulateThrows(coin map[string]float64, times int) map[string]int {

	// map to store the results of each toss
	results := make(map[string]int, times)

	for i := 0; i < times; i++ {

		// chooose function to select a side
		result := choose(coin)

		// check if the result has been obtained before
		i, exists := results[result]

		// if result obtained for the first time, initialize with zero
		if !exists {
			results[result] = 0
		}

		// increment the side by one for each face up on completion of throw
		results[result] = i + 1
	}

	return results
}

func main() {

	fmt.Println("===== Fair Dice =====")

	// unbiased dice
	coin := map[string]float64{
		"1": 1,
		"2": 1,
		"3": 1,
		"4": 1,
		"5": 1,
		"6": 1,
	}

	// create probability distribution
	normalizeMap(coin)

	// simulate throws
	results := simulateThrows(coin, 1000)

	// print the result
	fmt.Printf("Throw results: %v\n", results)

	fmt.Println("===== Loaded Dice =====")

	// biased dice
	weightedCoin := map[string]float64{
		"1": 10,
		"2": 1,
		"3": 1,
		"4": 1,
		"5": 1,
		"6": 1,
	}

	// create probability distribution
	normalizeMap(weightedCoin)

	// simulate throws
	results = simulateThrows(weightedCoin, 1000)

	// print the result
	fmt.Printf("Throw results: %v\n", results)
}

File: 03 - Probability/python/03_probability_solution.py
import random
from collections import OrderedDict


def normaliseSum(*args):
    """Returns a list of all the arguments, so that they add up to 1

    Args:
        *args: A variable number of arguments

    Returns:
        (list): list of all the arguments that all add up to 1
    """

    # convert the arguments to a list, so we can operate on it
    args = list(args)

    # calculate the sum of the list
    s = sum(args)

    # this if statement is important, because dividing by 0 is bad!
    if s != 0:
        # return a list on numbers divided by the sum
        return [float(x) / s for x in args]

    # return a list of 0s if the sum is 0
    return [0.0 for x in args]


def normaliseDict(distDict):
    """Returns a new dictionary where the values all add up to 1

    Args:
        distDict (dict): A probability distribution

    Returns:
        norm (dict): a new dictionary where the values all add up to 1
    """

    # get the keys from the disctionary
    keys = list(distDict.keys())

    # normalise the values from the dictionary
    norm = normaliseSum(*distDict.values())

    # create a normalised disctionary
    normalised = {}

    # iterate through the keys and values
    i = 0
    for key in keys:
        normalised[key] = norm[i]
        i += 1

    return normalised


def getCumulativeDistr(distr):
    """Returns an OrderedDictionary where the values add up the cumulative sum

    Args:
        distr (dict): A probability distribution

    Returns:
        norm (dict): a new dictionary where the values add up the cumulative sum
    """

    # first, make sure we have a normalised distribution
    distr = normaliseDict(distr)

    # create an OrderedDict
    cumulative = OrderedDict()

    # iterate through the distribution, adding up the probabilities as we go
    sum = 0
    for key in distr:
        sum += distr[key]
        cumulative[key] = sum

    # make sure the final key is a dead cert
    distr[key] = 1

    return cumulative


def sample(distr):
    """Picks a random value from a probability distribution

    Args:
        distr (dict): A probability distribution

    Returns:
        key (str): the chosen dict key
    """

    # make sure the distribution is a normalised cumulative distr
    distr = getCumulativeDistr(distr)

    # pick a random number between 0 and 1
    r = random.random()

    # iterate through the distribution
    for key in distr:
        # return the key that matches the random choice
        if r < distr[key]:
            return key

    # return the last key if all else fails
    return key


if __name__ == "__main__":

    print("=== Fair dice ===")

    fairDice = {
        "1": 1,
        "2": 1,
        "3": 1,
        "4": 1,
        "5": 1,
        "6": 1
    }

    results = {}
    for x in range(10):
        result = sample(fairDice)
        i = results.get(result, 0)
        results[result] = i + 1
    print("Throw results: {}".format(results))

    print("=== Loaded dice ===")

    loadedDice = {
        "1": 10,
        "2": 1,
        "3": 1,
        "4": 1,
        "5": 1,
        "6": 1
    }

    results = {}
    for x in range(10):
        result = sample(loadedDice)
        i = results.get(result, 0)
        results[result] = i + 1
    print("Throw results: {}".format(results))
